### PR TITLE
media-gfx/graphviz: add 4.0.0

### PR DESCRIPTION
**Pull request by thesamesam**
```
$ diff -ruN graphviz-2.50.0.ebuild graphviz-4.0.0.ebuild
--- graphviz-2.50.0.ebuild      2022-06-12 20:58:12.345907032 +0100
+++ graphviz-4.0.0.ebuild       2022-06-15 17:58:55.147125780 +0100
@@ -4,7 +4,7 @@
 EAPI=8

 PYTHON_COMPAT=( python3_{8..10} )
-inherit autotools flag-o-matic java-pkg-opt-2 python-single-r1 qmake-utils
+inherit flag-o-matic python-single-r1

 DESCRIPTION="Open Source Graph Visualization Software"
 HOMEPAGE="https://www.graphviz.org/ https://gitlab.com/graphviz/graphviz/"
@@ -14,8 +14,8 @@

 LICENSE="CPL-1.0"
 SLOT="0"
-KEYWORDS="~alpha amd64 ~arm arm64 hppa ~ia64 ~loong ~m68k ~mips ppc ppc64 ~riscv ~s390 sparc x86 ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~sparc-solaris ~x64-solaris"
-IUSE="+cairo devil doc examples gtk2 gts guile java lasi lefty nls pdf perl postscript python qt5 ruby svg tcl webp X"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~m68k ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86 ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~sparc-solaris ~x64-solaris"
+IUSE="+cairo devil doc examples gtk2 gts guile lasi nls pdf perl postscript python qt5 ruby svg tcl webp X"

 REQUIRED_USE="
        !cairo? ( !X !gtk2 !postscript !lasi )
@@ -32,9 +32,9 @@
        >=media-libs/fontconfig-2.3.95
        >=media-libs/freetype-2.1.10
        >=media-libs/gd-2.0.34:=[fontconfig,jpeg,png,truetype,zlib]
+       media-libs/libjpeg-turbo:=
        >=media-libs/libpng-1.2:0=
        sys-libs/zlib
-       virtual/jpeg:0
        virtual/libiconv
        cairo? (
                >=x11-libs/cairo-1.1.10[svg]
@@ -47,12 +47,7 @@
        )
        gts? ( sci-libs/gts )
        guile? ( dev-scheme/guile )
-       java? ( >=virtual/jre-1.8:* )
        lasi? ( media-libs/lasi )
-       lefty? (
-               x11-libs/libXaw
-               x11-libs/libXmu
-       )
        pdf? ( app-text/poppler )
        perl? ( dev-lang/perl:= )
        postscript? ( app-text/ghostscript-gpl )
@@ -69,7 +64,6 @@
        webp? ( media-libs/libwebp:= )
        X? ( x11-libs/libX11 )"
 DEPEND="${RDEPEND}
-       lefty? ( x11-base/xorg-proto )
        X? ( x11-base/xorg-proto )"
 BDEPEND="
        sys-devel/flex
@@ -83,10 +77,6 @@
                dev-lang/swig
                dev-scheme/guile
        )
-       java? (
-               dev-lang/swig
-               >=virtual/jdk-1.8:*
-       )
        nls? ( >=sys-devel/gettext-0.14.5 )
        perl? ( dev-lang/swig )
        python? ( dev-lang/swig )
@@ -104,8 +94,6 @@
 #   gtk-2 depends on pango, cairo and libX11 directly.
 # - gdk-pixbuf
 #   Directly depends on gtk-2 and gdk-pixbuf.
-# - glitz
-#   Disabled, no particular reason
 #   needs 'pangocairo' enabled in graphviz configuration
 # - ming
 #   flash plugin via -Tswf requires media-libs/ming-0.4. Disabled as it's
@@ -124,7 +112,6 @@
 # - scheme (enabled via guile) ... no longer broken on ~x86
 # - go (disabled)
 # - io (disabled)
-# - java (enabled via java) *2
 # - lua (enabled via lua)
 # - ocaml (enabled via ocaml)
 # - perl (enabled via perl) *1
@@ -141,10 +128,8 @@
 # tkspline, tkstubs ; enabled with: --with-tk

 # And the commands (/cmd):
-# - dot, dotty, gvedit, gvpr, lefty, lneato, smyrna, tools/* :)
+# - dot, gvedit, gvpr, smyrna, tools/* :)
 #   sci-libs/gts can be used for some of these
-# - lefty:
-#   needs Xaw and X to build
 # - gvedit (via 'qt5'):
 #   based on ./configure it needs qt-core and qt-gui only
 # - smyrna : experimental opengl front-end (via 'smyrna')
@@ -155,35 +140,6 @@

 pkg_setup() {
        use python && python-single-r1_pkg_setup
-
-       java-pkg-opt-2_pkg_setup
-}
-
-src_prepare() {
-       default
-
-       # ToDo: Do the same thing for examples and/or
-       #       write a patch for a configuration-option
-       #       and send it to upstream
-       # note - the longer sed expression removes multi-line assignments that are extended via '\'
-       if ! use doc ; then
-               find . -iname Makefile.am \
-                       | xargs sed -i -e '/^\(html\|pdf\)_DATA.*\\[[:space:]]*$/{:m;N;s/\\\n//;tm;d}' \
-                               -e '/^\(html\|pdf\)_DATA/d' || die
-       fi
-
-       # Nuke the dead symlinks for the bindings
-       sed -i -e '/$(pkgluadir)/d' tclpkg/gv/Makefile.am || die
-
-       # replace the whitespace with tabs
-       sed -i -e 's:  :\t:g' doc/info/Makefile.am || die
-
-       # use correct version of qmake. bug #567236
-       sed -i -e "/AC_CHECK_PROGS(QMAKE/a AC_SUBST(QMAKE,$(qt5_get_bindir)/qmake)" configure.ac || die
-
-       use java && append-cppflags $(java-pkg_get-jni-cflags)
-
-       eautoreconf
 }

 src_configure() {
@@ -192,6 +148,7 @@
                --cache-file="${S}"/config.cache
                --enable-ltdl
                $(use_with cairo pangocairo)
+               $(use_with examples demos '$(docdir)/examples')
                $(use_with devil)
                $(use_with gtk2 gdk)
                $(use_with gtk2 gdk-pixbuf)
@@ -204,7 +161,6 @@
                $(use_with svg rsvg)
                $(use_with webp)
                $(use_with X x)
-               $(use_enable lefty)
                --with-digcola
                --with-fontconfig
                --with-freetype2
@@ -213,13 +169,11 @@
                --with-sfdp
                --without-ming
                # new/experimental features, to be tested, disable for now
-               --without-glitz
                --without-ipsepcola
                --without-smyrna
                --without-visio
                # Bindings:
                $(use_enable guile)
-               $(use_enable java)
                $(use_enable perl)
                $(use_enable python python3)
                $(use_enable ruby)
@@ -227,6 +181,7 @@
                --disable-go
                --disable-io
                --disable-lua
+               --disable-java
                --disable-ocaml
                --disable-php
                --disable-python
@@ -240,32 +195,17 @@
 }

 src_install() {
-       sed -i -e "s:htmldir:htmlinfodir:g" doc/info/Makefile || die
-
-       emake DESTDIR="${D}" \
-               txtdir="${EPREFIX}"/usr/share/doc/${PF} \
-               htmldir="${EPREFIX}"/usr/share/doc/${PF}/html \
-               htmlinfodir="${EPREFIX}"/usr/share/doc/${PF}/html/info \
-               pdfdir="${EPREFIX}"/usr/share/doc/${PF}/pdf \
-               pkgconfigdir="${EPREFIX}"/usr/$(get_libdir)/pkgconfig \
-               install
-
-       if ! use examples; then
-               rm -rf "${ED}"/usr/share/graphviz/demo || die
-       fi
+       default

        find "${ED}" -name '*.la' -delete || die

-       einstalldocs
-
        use python && python_optimize \
                "${D}"$(python_get_sitedir) \
                "${ED}"/usr/$(get_libdir)/graphviz/python3
 }

 pkg_postinst() {
-       # This actually works if --enable-ltdl is passed
-       # to configure
+       # We need to register all plugins before they become usable
        dot -c || die
 }
```

### CODE REVIEW

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+# - smyrna : experimental opengl front-end (via 'smyrna')
+#   currently disabled -- it segfaults a lot
+#   needs x11-libs/gtkglext, gnome-base/libglade, media-libs/freeglut
+#   sci-libs/gts, x11-libs/gtk.  Also needs 'gtk','glade','glut','gts' and 'png'
+#   with flags enabled at configure time
+
+pkg_setup() {
+	use python && python-single-r1_pkg_setup
+
+	java-pkg-opt-2_pkg_setup
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:159]: get rid of java, don't care anymore

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+	# ToDo: Do the same thing for examples and/or
+	#       write a patch for a configuration-option
+	#       and send it to upstream
+	# note - the longer sed expression removes multi-line assignments that are extended via '\'
+	if ! use doc ; then
+		find . -iname Makefile.am \
+			| xargs sed -i -e '/^\(html\|pdf\)_DATA.*\\[[:space:]]*$/{:m;N;s/\\\n//;tm;d}' \
+				-e '/^\(html\|pdf\)_DATA/d' || die
+	fi
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:173]: this whole block is upstreamed, remove

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+	# Nuke the dead symlinks for the bindings
+	sed -i -e '/$(pkgluadir)/d' tclpkg/gv/Makefile.am || die
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:176]: remove

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+	use java && append-cppflags $(java-pkg_get-jni-cflags)
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:178]: remove, don't care about java anymore

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+	eautoreconf
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:180]: shouldn't be necessary anymore, not patching Makefile.am anymore

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+}
+
+src_configure() {
+	local myconf=(
+		# Speeds up the libltdl configure
+		--cache-file="${S}"/config.cache
+		--enable-ltdl
+		$(use_with cairo pangocairo)
+		$(use_with examples demos)
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:189]: append the standard examples dir as the third argument here in here
https://gitlab.com/graphviz/graphviz/-/merge_requests/2668
```suggestion
		$(use_with examples demos '$(docdir)/examples')
```

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	)
+	econf "${myconf[@]}"
+}
+
+src_install() {
+	default
+
+	find "${ED}" -name '*.la' -delete || die
+
+	einstalldocs
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:242]: shouldn't be necessary?

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+	use python && python_optimize \
+		"${D}"$(python_get_sitedir) \
+		"${ED}"/usr/$(get_libdir)/graphviz/python3
+}
+
+pkg_postinst() {
+	# This actually works if --enable-ltdl is passed
+	# to configure
+	dot -c || die
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:252]: nix the comment and replace it with
`# we need to register all plugins before they become usable`

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: media-gfx/graphviz/graphviz-4.0.0.ebuild (hunk 2 above)
Comment (thesamesam) [reply media-gfx/graphviz/graphviz-4.0.0.ebuild:173]: I thought I'd done those!

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+DESCRIPTION="Open Source Graph Visualization Software"
+HOMEPAGE="https://www.graphviz.org/ https://gitlab.com/graphviz/graphviz/"
+# unfortunately upstream uses an "artifact" store for the pre-generated
+# tarball now, which makes predictable URLs impossible.
+SRC_URI="https://gitlab.com/api/v4/projects/4207231/packages/generic/graphviz-releases/${PV}/${P}.tar.xz"
+
+LICENSE="CPL-1.0"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~m68k ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc ~x86 ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~sparc-solaris ~x64-solaris"
+IUSE="+cairo devil doc examples gtk2 gts guile lasi lefty nls pdf perl postscript python qt5 ruby svg tcl webp X"
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:18]:
```suggestion
IUSE="+cairo devil doc examples gtk2 gts guile lasi nls pdf perl postscript python qt5 ruby svg tcl webp X"
```
and remove everything dependent on it

**Review by SoapGentoo — CHANGES_REQUESTED** [review]

File: media-gfx/graphviz/graphviz-4.0.0.ebuild [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+RDEPEND="
+	>=dev-libs/expat-2
+	>=dev-libs/glib-2.11.1:2
+	dev-libs/libltdl:0
+	>=media-libs/fontconfig-2.3.95
+	>=media-libs/freetype-2.1.10
+	>=media-libs/gd-2.0.34:=[fontconfig,jpeg,png,truetype,zlib]
+	>=media-libs/libpng-1.2:0=
+	sys-libs/zlib
+	virtual/jpeg:0
```
Comment (SoapGentoo) [inline media-gfx/graphviz/graphviz-4.0.0.ebuild:37]: needs updating

**Review by SoapGentoo — APPROVED** [review]

**Pull request closed without merging**